### ISSUE TITLE
Support Console.beep on Linux (Fedora/RHEL)

### ISSUE THREAD

**Issue by Petermarcu** [dotnet/corefx#25746]
@chiddekel commented on [Tue Dec 05 2017](https://github.com/dotnet/core/issues/1130)

Run on OS: Fedora 27  
Kernel Linux:  4.13.16-302.fc27.x86_64  

Step zero: `yum install beep`  
First, run` sudo modprobe pcspkr` and then beep should work.
  
Additional uncomment the /etc/modprobe.d/beep.conf line at:
`# alias platform:pcspkr pcspkr`
to
`alias platform:pcspkr pcspkr`
For automatic load pcspkr module  

 Run application beep e.g -  `/usr/bin/beep  -f 2000 -l 1000 ` 
 The beep in this case equal to n = 2000 Hz and duration 1000 ms

This is tiny, small C source includes, of beep application. 
 https://goo.gl/E3kfPy  

The portion of this code can be re-write to C# and use in Console.beep implementation for Linux platform.

**Comment by stephentoub** [comment]
For clarification, `Console.Beep()` works fine on Unix.  It's `Console.Beep(int, int)` that throws PlatformNotSupportedException, due to lack of a good cross-platform mechanism for implementing it and it's relative value.

**Comment by chiddekel** [comment]
The implementation of platform Linux, Darwin, NT it simple if else statement or case switch.  
The code run after e.g If (platform == Linux) { } the detect it restricted only this way:  
  
The sounds are caused by the BIOS (Basic Input/Output System) by the OS (Operating System), by the DE (Desktop Environment).

**Comment by stephentoub** [comment]
You're welcome to submit a PR to be evaluated. I do not believe this is something folks at Microsoft will invest in any time soon.

**Comment by chiddekel** [comment]
In short - implementation it's not hard to implement.
All pice is in place.
  
Linux Developer or distro maintainer of eg, Fedora, Debian etc, etc  
Must check two-step:  
 1. Is module `pcspkr` un-blacklist  
 2. Is `PC Speaker, PC Beep, or Beep` unmute for example by alsa.

**Comment by ghuntley** [comment]
`Console.Beep(int,int)` on linux is indeed stubbed out. 

*NIX impl: https://github.com/dotnet/corefx/blob/master/src/System.Console/src/System/ConsolePal.Unix.cs#L199
Windows impl: https://github.com/dotnet/corefx/blob/master/src/System.Console/src/System/ConsolePal.Windows.cs#L679

**Comment by ghuntley** [comment]
Source code for `/usr/bin/beep` is at http://security.debian.org/debian-security/pool/updates/main/b/beep/beep_1.3.orig.tar.gz. 

Question for corefx folks:

> Is shelling out to `/usr/bin/beep` acceptable or would you prefer that 👇 is re-implemented.

![image](https://user-images.githubusercontent.com/127353/40815869-da65ee86-658c-11e8-9add-bc903ca7103a.png)

**Comment by danmosemsft** [comment]
@stephentoub thoughts? I'd prefer to port the code, as long as it's the same for all distro, since shelling out is not free (is another dependency, and is clumsy)

**Comment by stephentoub** [comment]
> I'd prefer to port the code, as long as it's the same for all distro, since shelling out is not free (is another dependency, and is clumsy)

I agree.  However:
a) What's the license on that code?  Porting something GPL for example would not be acceptable.
b) How much code is it?  The snippet shown above is calling methods (e.g. do_beep) where the src isn't shown.

It seems like an implementation should be possible, at least on Linux, with relatively little code, but if the cited source is GPL or a similar license, someone who's not looked at the code will need to provide the implementation.

**Comment by paule96** [comment]
Hi,

has anybody start the work on this?

**Comment by danmosemsft** [comment]
@paule96 I don't think so, you are welcome to offer a PR but note the comment above that it must be a clean implementation without consulting code under GPL or similar license.

**Comment by paule96** [comment]
I will try my best maybe :) But I don't have any know how about programming nativ stuff on Linux 😄 have anybody a good starting tutorial for this stuff? 🆘

**Comment by Cheesebaron** [comment]
On unix you open the console `/dev/tty0`, `/dev/vc/0`, `/dev/console` and write to it. However, as far as I know this requires root permissions. So writing directly to it will be a problem.

`beep` usually has its user id set to root.

**Comment by Cheesebaron** [comment]
The function you have found is using `dos.h`, that doesn't seem like right place to look.

**Comment by chiddekel** [comment]
Link was unavailable - here is re paste Debian git repo. http://git.deb.at/w/pkg/beep.git

**Comment by stephentoub** [comment]
> here is re paste Debian git repo

That code is GPL.

**Comment by chiddekel** [comment]
Yes code is GPL - re writed code based on beep Debian source code 👆- should be include the rule of interoperable in copyright law adding portion of code under GPL licence.

**Comment by stephentoub** [comment]
> should be include the rule of interoperable in copyright law adding portion of code under GPL licence.

Sorry, I'm not quite understanding, but we _cannot_ base code in this repo on anything that is GPL.

**Comment by chiddekel** [comment]
You can

**Comment by chiddekel** [comment]
You can, new portion of code will be under GPL licence, and licence can be add to this specific portion of code, so will be include only to beep re-implementation.

**Comment by stephentoub** [comment]
No, that is not how this repo / product operates.
